rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -125,6 +125,89 @@ object Reduce {
     )(implicit costAccountingAlg: CostAccountingAlg[M]): M[Unit] =
       tuplespaceAlg.consume(binds, ParWithRandom(body, rand), persistent)
 
+    private trait EvalJob {
+      def run(starts: Vector[Int], startIdx: Int)(
+          env: Env[Par],
+          rand: Blake2b512Random,
+          costAccountingAlg: CostAccountingAlg[M]
+      ): M[List[Unit]]
+      def size: Int
+    }
+
+    private object EvalJob {
+
+      private def split(
+          starts: Vector[Int],
+          rand: Blake2b512Random,
+          startIdx: Int,
+          idx: Int
+      ): Blake2b512Random =
+        if (starts.last == 1)
+          rand
+        else if (starts.last > 256)
+          rand.splitShort((starts(startIdx) + idx).toShort)
+        else
+          rand.splitByte((starts(startIdx) + idx).toByte)
+
+      private def mkJob[A](
+          input: Seq[A],
+          handler: A => (Env[Par], Blake2b512Random, CostAccountingAlg[M]) => M[Unit]
+      ): EvalJob =
+        new EvalJob() {
+          override def run(starts: Vector[Int], startIdx: Int)(
```

Comment:
Previous implementation, when starting the evaluation of certain kind of terms didn't need to know about all start indexes (`starts`). Single `run` call on specific type of terms should only know about it's own index. It means that splitting needs to be done outside of `run`. Your `split` methods splits at `starts(startIdx) + idx` so it can be improved even more.

Reply:
If you compare to the original implementation, the "rand splitting code" was referring start indices. Indeed, it might be a good idea to split the rngs upfront and zip them with the jobs